### ISSUE TITLE
[Package Request]: PHP & Composer

### ISSUE THREAD

**Issue by GamerClassN7** [microsoft/winget-pkgs#56226]
### Package Requested

- [X] I would like someone else to build the manifest.
- [ ] I would like help so I can submit the manifest.
- [X] I have performed a search and couldn't find this package.
- [ ] I have checked that the installer for this package is not a standalone or portable executable or a .zip file
- [ ] I think there is a new version available and I have provided the URL.

### Please provide the following information

Publisher: [The PHP Group](https://www.php.net/copyright.php)
Publisher Url: https://www.php.net/
Package Name:  PHP
Description: PHP Executables for local usage and development
Package Version:  8.1.4
Installer URL: https://windows.php.net/downloads/releases/php-8.1.4-Win32-vs16-x64.zip

* Also if possible also implement composer ass well

Publisher: [Nils Adermann](http://naderman.de/), [Jordi Boggiano](https://seld.be/) and many [community contributions](https://github.com/composer/composer/graphs/contributors)
Publisher Url: https://getcomposer.org/
Package Name:  Composer
Description: PHP Dependeancy Manager
Package Version:  2.2.10
Installer URL: https://getcomposer.org/Composer-Setup.exe

**Comment by Trenly** [comment]
Blocked by microsoft/winget-cli#140

**Comment by Kasparsu** [comment]
Can somebody fix the title? Its composer not compooser

**Comment by GamerClassN7** [comment]
> Can somebody fix the title? Its composer not compooser

Done

**Comment by mdanish-kh** [comment]
Composer cannot be added because WinGet does not support installing dependencies yet.
- https://github.com/microsoft/winget-pkgs/pull/58422

As for adding PHP exectuable, it is blocked by
- https://github.com/microsoft/winget-cli/issues/2711

**Comment by hueldoeu** [comment]
i know this is an unpopular opinion but i want to say it. i know that you can install microsoft iis easily on windows 11 but if there is no cmd => winget install php, then cmd => winget install ApacheFriends.Xampp.X.X is the only easiest way to start a webserver on your computer.

**Comment by Kasparsu** [comment]
> i know this is an unpopular opinion but i want to say it. i know that you can install microsoft iis easily on windows 11 but if there is no cmd => winget install php, then cmd => winget install ApacheFriends.Xampp.X.X is the only easiest way to start a webserver on your computer.

yes very unpopular since it installs lot of bloat and has horrible defaults. Use scoop or chocolatey in meanwhile

**Comment by Trenly** [comment]
[Policy] Unblocked; Will require `ManifestVersion: 1.9.0` with `ArchiveBinariesDependOnPath: true`

**Comment by nbro** [comment]
We can already install PHP with winget, but we can't yet install `composer`. What's preventing people from adding support to install composer? If you want, I could try to do it. It doesn't seem so complicated.

It would also be great if we could somehow also be able to install PEAR and PECL when installing PHP, although I am not sure if these tools are compatible with `winget`, but being able to install/uninstall them with `winget` automatically would be great.

**Comment by lt-mat** [comment]
Hi! Is there any update on this? It would be really useful for many of us!

**Comment by Dragon1573** [comment]
> Hi! Is there any update on this? It would be really useful for many of us!

There's a <https://github.com/microsoft/winget-pkgs/labels/Scripted-Application> label attached to this issue.

`composer` and other related stuffs are Batch files. They're currently NOT supported by `winget.exe`.

- See microsoft/winget-cli#5085
- See microsoft/winget-cli#5083

**Comment by DandelionSprout** [comment]
Testing Composer now:

<img width="1234" height="246" alt="Image" src="https://github.com/user-attachments/assets/c666dd21-d3fd-4002-956a-ef457453940c" />

Looking through the test's generated log file, the line in it that stood out to me was `$ Error, no php selected`.

**Comment by Kasparsu** [comment]
<img width="962" height="492" alt="Image" src="https://github.com/user-attachments/assets/733b6e85-03ba-4b87-8f6e-aa3245ab072c" />

<img width="1206" height="115" alt="Image" src="https://github.com/user-attachments/assets/0fbc82bf-8a0c-45aa-b4d9-e5920482f5c6" />

Seems to be working for me ;P you need to have PHP installed first `winget install PHP.PHP.8.4`

**Comment by Dragon1573** [comment]
> Seems to be working for me ;P you need to have PHP installed first `winget install PHP.PHP.8.4`

Nice 🎉 

But you just lock dependencies down to PHP `v8.4` series. Composer might not only work for this specific version.

`winget.exe` does not support "or" or globs for dependency package identifiers. If a user already installed other valid PHP instance, they might facing a duplicate PHP installation because of the hard-coded package dependencies. 😞

**Comment by Kasparsu** [comment]
> > Seems to be working for me ;P you need to have PHP installed first `winget install PHP.PHP.8.4`
> 
> Nice 🎉
> 
> But you just lock dependencies down to PHP `v8.4` series. Composer might not only work for this specific version.
> 
> `winget.exe` does not support "or" or globs for dependency package identifiers. If a user already installed other valid PHP instance, they might facing a duplicate PHP installation because of the hard-coded package dependencies. 😞

I did not make or modify the manifest files just testing if it works. I agree that composer should not depend on php version. Also it is kinda weird that we have different php versions as separate packages instead as single package with different versions. If I remember correctly you can do that in winget as well.

**Comment by thiagodp** [comment]
Howdy! 
**Any updates on this?** 

It is currently possible to install Composer from a Manifest file? How/Where I can find it?
(I couldn't find it at https://github.com/microsoft/winget-pkgs/tree/master/manifests/c)

**Comment by Dragon1573** [comment]
> It is currently possible to install Composer from a Manifest file? How/Where I can find it?

**Nope.**

- Blocked by https://github.com/microsoft/winget-pkgs/issues/56226#issuecomment-3261699524

**Comment by thiagodp** [comment]
@Dragon1573 What is needed to unblock it?

**Comment by Dragon1573** [comment]
> What is needed ...

I think a new feature of `winget.exe` is rrquired, to support wildcard package identifiers (like `PHP.PHP.*` represents all valid PHP versions) or "selectable" dependency group (like `[PHP.PHP.8.*, PHP.PHP.7.*, PHP.PHP.6.*]` locks required PHP version from v6 to v8).

**Comment by thiagodp** [comment]
@Dragon1573 Thank you. So if I'm not mistaken, we could have different Composer packages for each PHP version, right? Something like `Composer.PHP.8.4`, `Composer.PHP.8.5`. That solution could be used while Winget does not offer wildcard support.

**Comment by DandelionSprout** [comment]
Some complications have emerged, which would be inconvenient at best, and blockage causes at worst.

Officially and at least partially (though not consistently IIRC) enforced, is that this repo prohibits what are deemed "Scripted applications", which stephengillie's research in https://github.com/microsoft/winget-pkgs/pull/360657#issuecomment-4261362812 indicate could possibly be the case. I presume a lot of PHP users know more about that than what me and he do.

A considerably bigger problem in my eyes is that Composer's official .exe is not version-specific, as it has used the same .exe since 2021 (6.3.0 according to https://github.com/composer/windows-setup). The result is that when a new version of Composer (e.g. 2.9.7) gets released upstream, the corresponding Windows installer .exe remains at the same file version, installer URL, and SHA256 hash. This makes it next-to-impossible for Winget's package updater third-party bots to determine when a new Composer version has been released.

Manual manifest updates would be possible, but that could also get strange since each new manifest update would be the exact same .exe but with a higher version number.

**Comment by DandelionSprout** [comment]
I at least got it added to the third-party repo "[Winget-extras](https://github.com/pl4nty/winget-extras)" now: `winget source add --name winget-extras --type Microsoft.PreIndexed.Package --arg https://winget.tplant.com.au/cache; winget install Composer.WindowsSetup`
